dist/tools/usb-serial/ttys.py

```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/python3
+"""
+Command line utility to list and filter TTYs
+"""
+import argparse
+import json
+import os
+import sys
+import time
+
+import pyudev
+
+
+def unescape(string):
+    """
+    Decodes unicode escaping in a string, e.g. "Hallo\\x20World" is decoded as
+    "Hallo World"
+    """
+    return bytes(string, "utf-8").decode("unicode_escape")
```

Comment:
```suggestion
    return bytes(string, "ascii").decode("unicode_escape").encode('latin1').decode('utf8', error='replace')
```

The 'utf-8' in the original version is a red herring: the input is ASCII per the (presumed) escaping rules. Also the "unicode_escape" is misleading (but Python has no equivalent "c_escape"): It would properly unescape any `\u1234' strings, but what's actually there is `\x89`, and that's part of Python's still weird admittance of latin1's special position in the encoding world that got grandfathered into Unicode by the assignment of code points 128-255.

That all result is the mojibake you see on Universit*t.

The proposed changes are still lossy (a bit like the underscores), but in a better defined way (precisely those characters that are not valid UTF-8 are replaced), and still easy to enter (you can enter regular text, or copy-paste the output around).

This loses bytes that are not UTF-8, which could in theory be salvaged through surrogate escaping, but if someone's device is named "foo\xffbar", that's not because their whole system is using a codec where \xff is valid, but just because someone screwed up in firmware development. (Not that there'd be a rule that these things are UTF-8, but anything else is just so 20th century).

Reply:
This gives me

```
$ make list-ttys
Traceback (most recent call last):
  File "/home/maribu/Repos/software/RIOT/dist/tools/usb-serial/ttys.py", line 220, in <module>
    print_ttys(sys.argv)
  File "/home/maribu/Repos/software/RIOT/dist/tools/usb-serial/ttys.py", line 205, in print_ttys
    tty = tty2dict(dev)
  File "/home/maribu/Repos/software/RIOT/dist/tools/usb-serial/ttys.py", line 36, in tty2dict
    result["vendor"] = unescape(dev.get("ID_VENDOR_ENC"))
  File "/home/maribu/Repos/software/RIOT/dist/tools/usb-serial/ttys.py", line 19, in unescape
    res = bytes(string, "ascii").decode("unicode_escape")
UnicodeEncodeError: 'ascii' codec can't encode character '\xe4' in position 18: ordinal not in range(128)
make: *** [/home/maribu/Repos/software/RIOT/tests/periph_timer_periodic/../../Makefile.include:867: list-ttys] Error 1
```

Using `bytes(string, "utf8")` instead of `bytes(string, "ascii")` gives me:

```
$ make list-ttys
path         | driver   | vendor                   | model  | model_db               | serial   | ctime   
-------------|----------|--------------------------|--------|------------------------|----------|---------
/dev/ttyUSB0 | ftdi_sio | Freie Universität Berlin | MSB-A2 | FT232 Serial (UART) IC | ARR6DKEE | 20:21:22
```

Reply:
That's

* funny because it means that udev is letting high bytes go through unescaped (while escaping even whitespace with \x20)
* weird because I've just tested it and when I enter UTF-8 in a RIOT USB device's name that conversion makes it break ...

... except that I just found that USB identifiers are UTF-16 (or UCS-2?); reading up a little to ensure that we don't nail down a conversion here that turns out to be broken just because it happens to work at the Universität device.

Reply:
The Umlaut actually looks pretty UTF-8 like to me:

```
$ udevadm info /dev/ttyUSB0 
P: /devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/ttyUSB0/tty/ttyUSB0
N: ttyUSB0
S: serial/by-id/usb-Freie_Universität_Berlin_MSB-A2_ARR6DKEE-if00-port0
S: serial/by-path/pci-0000:00:14.0-usb-0:2:1.0-port0
E: DEVLINKS=/dev/serial/by-id/usb-Freie_Universität_Berlin_MSB-A2_ARR6DKEE-if00-port0 /dev/serial/by-path/pci-0000:00:14.0-usb-0:2:1.0-port0
E: DEVNAME=/dev/ttyUSB0
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/ttyUSB0/tty/ttyUSB0
E: ID_BUS=usb
E: ID_MODEL=MSB-A2
E: ID_MODEL_ENC=MSB-A2
E: ID_MODEL_FROM_DATABASE=FT232 Serial (UART) IC
E: ID_MODEL_ID=6001
E: ID_PATH=pci-0000:00:14.0-usb-0:2:1.0
E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_2_1_0
E: ID_PCI_CLASS_FROM_DATABASE=Serial bus controller
E: ID_PCI_INTERFACE_FROM_DATABASE=XHCI
E: ID_PCI_SUBCLASS_FROM_DATABASE=USB controller
E: ID_REVISION=0600
E: ID_SERIAL=Freie_Universität_Berlin_MSB-A2_ARR6DKEE
E: ID_SERIAL_SHORT=ARR6DKEE
E: ID_TYPE=generic
E: ID_USB_DRIVER=ftdi_sio
E: ID_USB_INTERFACES=:ffffff:
E: ID_USB_INTERFACE_NUM=00
E: ID_VENDOR=Freie_Universität_Berlin
E: ID_VENDOR_ENC=Freie\x20Universität\x20Berlin
E: ID_VENDOR_FROM_DATABASE=Future Technology Devices International, Ltd
E: ID_VENDOR_ID=0403
E: MAJOR=188
E: MINOR=0
E: SUBSYSTEM=tty
E: USEC_INITIALIZED=7836843200
```

Using

```python
def unescape(string):
    """
    Decodes unicode escaping in a string, e.g. "Hallo\\x20World" is decoded as
    "Hallo World"
    """
    return string.replace("\\x20", " ")
```

Yields:

```
path         | driver   | vendor                   | model  | model_db               | serial   | ctime   
-------------|----------|--------------------------|--------|------------------------|----------|---------
/dev/ttyUSB0 | ftdi_sio | Freie Universität Berlin | MSB-A2 | FT232 Serial (UART) IC | ARR6DKEE | 21:17:33
```

If udev really only replaces spaces and invalid stuff, restoring spaces would be the only thing needed. Maybe I should find some documentation on this :)

Reply:
I've verified that the current conversion produces correct output even when non-UCS2 model names are used; furthermore, the output is consistent with the output of lsusb.

The incantation looks weird in terms of locales, but that's just because Python's decode('unicode_escape') makes a hardcoded choice of latin1 for bytes that are present in the string in unescaped form, and that gets reverted in the latin1 encode step.

I think this is fine (but will need to make up my mind about whether I want to clean up the crude UTF-16 support I hacked into _cpy_str_to_utf16 for inclusion in RIOT :-) ).